src/main-process/comfyInstallation.ts

```diff
@@ -15,64 +14,158 @@ type InstallState = Exclude<DesktopSettings['installState'], undefined>;
+   * @param data The data to send to the renderer
+   */
+  onUpdate?: (data: InstallValidation) => void;
+
   constructor(
     /** Installation state, e.g. `started`, `installed`.  See {@link DesktopSettings}. */
     public state: InstallState,
     /** The base path of the desktop app.  Models, nodes, and configuration are saved here by default. */
-    public basePath: string
-  ) {}
+    basePath: string,
+    public device: TorchDeviceType | undefined
+  ) {
+    // TypeScript workaround: duplication of basePath setter
+    this.#basePath = basePath;
+    this.virtualEnvironment = new VirtualEnvironment(basePath, this.device);
+  }
 
   /**
    * Static factory method. Creates a ComfyInstallation object if previously saved config can be read.
    * @returns A ComfyInstallation (not validated) object if config is saved, otherwise `undefined`.
+   * @throws If YAML config is unreadable due to access restrictions
    */
   static fromConfig(): ComfyInstallation | undefined {
     const config = useDesktopConfig();
     const state = config.get('installState');
     const basePath = config.get('basePath');
-    if (state && basePath) return new ComfyInstallation(state, basePath);
+    const device = config.get('selectedDevice');
+    if (state && basePath) return new ComfyInstallation(state, basePath, device);
   }
 
   /**
-   * Validate the installation and add any results to {@link issues}.
-   * @returns The validated installation state, along with a list of any issues detected.
+   * Validate the installation and add any results to {@link ComfyInstallation.validation}.
+   * @returns The validated installation state
+   * @throws When the YAML file is present but not readable (access denied, FS error, etc).
    */
   async validate(): Promise<InstallState> {
     log.info(`Validating installation. Recorded state: [${this.state}]`);
-
-    let { state } = this;
+    const validation: InstallValidation = {
+      inProgress: true,
+      installState: this.state,
+    };
+    this.validation = validation;
+    const update = () => this.onUpdate?.(validation);
+    update();
 
     // Upgraded from a version prior to 0.3.18
     // TODO: Validate more than just the existence of one file
-    if (!state && ComfyServerConfig.exists()) {
+    if (!validation.installState && ComfyServerConfig.exists()) {
       log.info('Found extra_models_config.yaml but no recorded state - assuming upgrade from <= 0.3.18');
-      state = 'upgraded';
+      validation.installState = 'upgraded';
+      update();
     }
 
     // Validate base path
     const basePath = await this.loadBasePath();
-    if (basePath === undefined || !(await pathAccessible(basePath))) {
-      log.warn('"base_path" is inaccessible or undefined.');
-      this.issues.add('invalidBasePath');
+    if (basePath && (await pathAccessible(basePath))) {
+      validation.basePath = 'OK';
+      update();
+
+      // TODO: Validate content of venv, etc.
```

Comment:
Can you file an issue tracking this TODO, and add the issue URL here? The issue should explain what needs to be done in detail.